### ISSUE TITLE
[Linux] Cache path should not default to /tmp

### ISSUE THREAD

**Issue by yoshiyoshyosh** [Yellow-Dog-Man/Resonite-Issues#54]
### Describe the bug?

On Linux, the asset cache path defaults to `/tmp/Yellow Dog Man Studios/Resonite`. On almost all Linux distributions, `/tmp` is mounted as a `tmpfs`, meaning that the files in there exist in ram. Due to this, asset cache runs into two major problems:

1. It uses more memory on a system than needed
2. It does not persist after a computer restart, thus requiring assets to be fetched again

### To Reproduce

1. Do not use the `-CachePath` option when launching Resonite
2. Observe that, on launch, asset cache populates `/tmp/Yellow Dog Man Studios/Resonite`

### Expected behavior

The cache path is set to a reasonable, persistent location by default.

Optimally, the path for cache files would follow the [XDG Base Directory Specification](https://wiki.archlinux.org/title/XDG_Base_Directory), in which the cache path will default to `$XDG_CACHE_HOME/Yellow Dog Man Studios/Resonite` or, if the environment variable `XDG_CACHE_HOME` does not exist, `$HOME/.cache/Yellow Dog Man Studios/Resonite`

### Resonite Version Number

2023.10.13.743

### What Platforms does this occur on?

Linux

### Log Files

[Resonite log](https://gist.github.com/yoshiyoshyosh/aa26de760e3a1c53427d7a7dd3d4c06d)

*For ease on my part, this log file is reused from another bug report, but said bug report was only in home for a brief moment.*

### Additional Context

A current workaround for the issue is adding the launch option `-CachePath "${XDG_CACHE_HOME:-"$HOME/.cache"}/Yellow Dog Man Studios/Resonite"`

### Reporters

yosh

**Comment by Frooxius** [comment]
This is a little bit interesting, because for determining the default cache path, we are using Unity's temporary cache path: https://docs.unity3d.com/ScriptReference/Application-temporaryCachePath.html

So the path this one provides seems to exist only in RAM?

I'm not familiar enough with Linux, so I don't know what would be a good programmatic way to determine the default path within a Mono process (specifically running within Unity) that'll be generally compatible with distributions.

**Comment by iamgreaser** [comment]
On pretty much every modern Linux distro these days, `/tmp` points to RAM, yes. In most cases that would be fine, but in the case of the Resonite asset cache, it really is *too* temporary for this purpose.

Some pseudocode which should implement what is required (this snippet I release into the public domain):

```
xdg_cache = getenv("XDG_CACHE_HOME");
if (xdg_cache == null or xdg_cache == "") {
  home = getenv("HOME");
  if (home == null or home == "") { complain(); xdg_cache = "/tmp"; } // Unity would probably balk anyway, but just in case...
  else { xdg_cache = join_paths(home, ".cache"); }
}
cache_path = join_paths(xdg_cache, "Resonite");
```

Replace `getenv` with whatever you use to read environment variables, and replace `join_paths` with whatever .NET provides for joining paths in an OS-independent manner because it's a good habit to have. And if your `getenv` API throws an exception if something is missing, then adjust the code accordingly.

**Comment by Frooxius** [comment]
Thank you! Just to make sure I'm understanding things right, getenv() gets an environment variable?

So I essentially get that environment variable. If that gives nothing, I try to get path from HOME environment variable, adding `.cache` at the end. If that gives nothing, just fallback to /tmp?

I can get this implemented in.

**Comment by Banane9** [comment]
This stackoverflow question seems like it could help too:  
https://stackoverflow.com/questions/1143706/getting-the-path-of-the-home-directory-in-c

**Comment by yoshiyoshyosh** [comment]
> Thank you! Just to make sure I'm understanding things right, getenv() gets an environment variable?

seems to be so

> So I essentially get that environment variable. If that gives nothing, I try to get path from HOME environment variable, adding `.cache` at the end. If that gives nothing, just fallback to /tmp?

basically, yeah. the logic should be:
- get the environment variable `$XDG_CACHE_HOME`. if it exists, set the cache path to `$XDG_CACHE_HOME/Yellow Dog Man Studios/Resonite` (or just `$XDG_CACHE_HOME/Resonite`, I guess. depends on if you wanna be consistent cross-platform)
- if the environment variable is not set or is null, set the cache path to `$HOME/.cache/Yellow Dog Man Studios/Resonite`, creating parent folders as needed
  - `$HOME` should always exist when someone is logged in and playing resonite. I guess the link banane9 sent would be better, since it's an api directly within mono.
 - if all else fails (which should be incredibly, incredibly rare), just use `/tmp` as before

> I can get this implemented in.
:tada:

**Comment by Enverex** [comment]
> 
> I'm not familiar enough with Linux, so I don't know what would be a good programmatic way to determine the default path within a Mono process (specifically running within Unity) that'll be generally compatible with distributions.

So Linux follows some standard norms when it comes to content caching. The general rules are as follows:

~/.cache/(application name) : Persistent cache for this application.
/var/tmp : Persistent cache for global use.
/tmp : Non-persistent cache that should be cleared on reboot and on most systems is a RAM drive.

So even if /tmp wasn't a RAM drive, it's expected to be cleared on every reboot anyway which wouldn't work well as a cache for Resonite.

So as Yoshi said above, the most logical one to use would be the user home folder cache. That also avoids any potential permission issues as well.

**Comment by Frooxius** [comment]
I've implemented this in 2023.11.8.335, if there are still issues reopen. Thanks!
